### PR TITLE
Scripts for defect 3136

### PR DESCRIPTION
**Pull request by dboltovskyi**
ATF Test Scripts to check [#3136](https://github.com/smartdevicelink/sdl_core/issues/3136)

This PR is **ready** for review.

### Summary
Current PR has 2 types of changes
 - new scripts to check SDL issue
 - updates in existing script due to changes in SDL functionality

### ATF version
develop

### Changelog
 - added new test scripts/test sets
 - updated existing scripts
    - Policies (incl. update to speed up scripts for retry sequence)
    - Remote control feature
    - TheSameApp feature

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/Defects/6_1/3136_01_Proprietary.lua [hunk 1]
```diff
@@ -0,0 +1,148 @@
+---------------------------------------------------------------------------------------------------
+-- Issue: https://github.com/smartdevicelink/sdl_core/issues/3136
+--
+-- Description: Successful 2nd HMI PTU if it's triggered within failed retry for the 1st PTU
```
Comment (GetmanetsIrina) [inline test_scripts/Defects/6_1/3136_01_Proprietary.lua]: I think that `HMI PTU` should not be mentioned in scripts for develop.

File: test_scripts/TheSameApp/Policies/AppId/002_PTU_is_triggered_after_registration_app_with_different_app_id.lua [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+--[[ Scenario ]]
+runner.Title("Preconditions")
+runner.Step("Clean environment", common.preconditions)
+runner.Step("Start SDL and HMI", common.start)
+runner.Step("Connect two mobile devices to SDL", common.connectMobDevices, {devices})
+
+runner.Title("Test")
+runner.Step("Register App1 from device 1", common.registerAppWithPTU, { 1, appParams[1], 1 })
+runner.Step("Register App2 from device 2", common.registerAppWithoutPTU, { 2, appParams[2], 2 })
```
Comment (GetmanetsIrina) [inline test_scripts/TheSameApp/Policies/AppId/002_PTU_is_triggered_after_registration_app_with_different_app_id.lua]: Here `Check: SDL does not send SDL.OnStatusUpdate and BC.PolicyUpdate to HMI during the app registration` check for `OnStatusUpdate` is absent.

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+-- the first PTU for App1 from Mobile №1 is finished in case App1 and App2 are registered
+-- within 'ApplicationListUpdateTimeout'
+--
+-- Preconditions:
+-- 1) SDL and HMI are started
+-- 2) Mobile №1 and №2 are connected to SDL and are consented
+-- 3) App1 is registered from Mobile №1 and triggers PTU
+--
+-- Steps:
+-- 1) App2 is registered from Mobile №2 during PTU for App1 from Mobile №1 is in progress
```
Comment (GetmanetsIrina) [inline test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: App 2 is registered from Mobile 2 within ApplicationListUpdateTimeout during ...

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+--   Check: SDL does not send SDL.OnStatusUpdate and BC.PolicyUpdate to HMI during the app registration
+-- 2) First PTU is performed successful
+--   Check: SDL triggers new PTU after the first one is finished
+---------------------------------------------------------------------------------------------------
```
Comment (GetmanetsIrina) [inline test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: SDL should not trigger new PTU in case second app was registered within 'ApplicationListUpdateTimeout

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+end
+
+--[[ Scenario ]]
+runner.Title("Preconditions")
+runner.Step("Clean environment", common.preconditions)
+runner.Step("Start SDL and HMI", common.start)
+runner.Step("Connect two mobile devices to SDL", common.connectMobDevices, {devices})
+
+runner.Title("Test")
+runner.Step("Register App1 from device 1 - No PTU", registerApp1)
```
Comment (GetmanetsIrina) [inline test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: Propose to update the description to`Register App1 from device 1 - No PTU before ApplicationListUpdateTimeout is expired `

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+runner.Step("Register App2 from device 2 - PTU started", registerApp2)
```
Comment (GetmanetsIrina) [inline test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: -> "Register App2 from device 2 - PTU started after ApplicationListUpdateTimeout is expired"

File: test_scripts/TheSameApp/Policies/AppId/002_PTU_is_triggered_after_registration_app_with_different_app_id.lua (hunk 2 above)
Comment (dboltovskyi) [reply test_scripts/TheSameApp/Policies/AppId/002_PTU_is_triggered_after_registration_app_with_different_app_id.lua]: @GetmanetsIrina There is a check in `common.registerAppWithoutPTU()`:
```
common.hmi.getConnection():ExpectRequest("SDL.OnStatusUpdate"):Times(0)
```

File: test_scripts/Defects/6_1/3136_01_Proprietary.lua (hunk 1 above)
Comment (dboltovskyi) [reply test_scripts/Defects/6_1/3136_01_Proprietary.lua]: @GetmanetsIrina Please find update in 0e0c08b

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua (hunk 3 above)
Comment (dboltovskyi) [reply test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: @GetmanetsIrina Please find update in 0e0c08b

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua (hunk 4 above)
Comment (dboltovskyi) [reply test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: @GetmanetsIrina Please find update in 0e0c08b

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua (hunk 5 above)
Comment (dboltovskyi) [reply test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: @GetmanetsIrina Please find update in 0e0c08b

File: test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua (hunk 6 above)
Comment (dboltovskyi) [reply test_scripts/TheSameApp/Policies/AppId/005_PTU_is_not_triggered_after_registration_app_with_different_app_id_during_timeout.lua]: @GetmanetsIrina Please find update in 0e0c08b

**Comment by dboltovskyi** [comment]
@theresalech Please notice PR is ready for PM review

File: test_scripts/RC/OnRCStatus/012_notification_by_revoking_module_by_policy.lua [hunk 7]
```diff
@@ -51,6 +51,7 @@ end
 --[[ Scenario ]]
 runner.Title("Preconditions")
 runner.Step("Clean environment", common.preconditions, { true, 1 })
+runner.Step("Update SDL config", common.setSDLIniParameter, { "ApplicationListUpdateTimeout", 4000 })
```
Comment (ShobhitAd) [inline test_scripts/RC/OnRCStatus/012_notification_by_revoking_module_by_policy.lua]: Could you please explain why this change is necessary. This test passes for me even without these changes

File: test_scripts/TheSameApp/Policies/AppId/001_PTU_is_not_triggered_after_registration_app_with_the_same_id.lua [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+--
+-- Preconditions:
+-- 1) SDL and HMI are started
+-- 2) Mobile №1 and №2 are connected to SDL and are consented
+-- 3) App1 is registered from Mobile №1 and triggers PTU
+--
+-- Steps:
+-- 1) App1 is registered from Mobile №2 during PTU for App1 from Mobile №1 is in progress
+--   Check: SDL does not send SDL.OnStatusUpdate and BC.PolicyUpdate to HMI during the app registration
+-- 2) First PTU is performed successful
```
Comment (ShobhitAd) [inline test_scripts/TheSameApp/Policies/AppId/001_PTU_is_not_triggered_after_registration_app_with_the_same_id.lua]:
```suggestion
-- 2) First PTU is successful
```

File: test_scripts/Defects/6_1/3136_01_Proprietary.lua [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+  local str = "[" .. atf_logger.formated_time(true) .. "]"
+  for i, p in pairs({...}) do
+    local delimiter = "\t"
+    if i == 1 then delimiter = " " end
+    str = str .. delimiter .. p
+  end
+  utils.cprint(35, str)
+end
+
+local function preloadedTupd(pTbl)
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/Defects/6_1/3136_01_Proprietary.lua]: @ShobhitAd Please find update in e864083f6

File: test_scripts/Defects/6_1/3136_01_Proprietary.lua [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+--[[ Local Functions ]]
+local function log(...)
+  local str = "[" .. atf_logger.formated_time(true) .. "]"
+  for i, p in pairs({...}) do
+    local delimiter = "\t"
+    if i == 1 then delimiter = " " end
+    str = str .. delimiter .. p
+  end
+  utils.cprint(35, str)
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/Defects/6_1/3136_01_Proprietary.lua]: @ShobhitAd Please find update in e864083f6

File: test_scripts/RC/OnRCStatus/012_notification_by_revoking_module_by_policy.lua (hunk 7 above)
Comment (dboltovskyi) [reply test_scripts/RC/OnRCStatus/012_notification_by_revoking_module_by_policy.lua]: @ShobhitAd Sure.
SDL has a timeout defined by `ApplicationListUpdateTimeout` configuration option.
Once new app is registered SDL will wait during this timeout if other applications are registered as well.
Once this timeout expires SDL will send `BC.UpdateAppList` request to HMI and starts PTU sequence.
In this way it's possible to receive a PT update for a set of applications at once and not to have a few PTU sequences for each app.
By default this timeout is 3000 ms. and it may be not enough for our case:
 - Registering of 1st app takes 2000ms
 - Activation of 1st app - 500ms
 - Allocation of a new module - 100ms

We have a window of 300-400ms to register 2nd app and not to have a separate PTU sequence for it.
So we just give SDL more time to wait for the 2nd app in order to have one PTU for both apps.

**Comment by ShobhitAd** [comment]
@dboltovskyi Approved the changes. Something to note before you resolve the merge conflicts. We would appreciate if developers did not force push to a pr that is in review. It makes it difficult to review a PR because we cannot tell what was changed since the last review. In general, please do not rebase, do not force push against sdl_atf_test_scripts. Instead, you can merge the current develop branch into your PR branch. Thank you

**Comment by ShobhitAd** [comment]
@dboltovskyi The hmi PTU feature has been merged into develop as well. Once the merge conflicts are resolved, tests `016` and `017` in https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_sets/hmi_ptu.txt#L16 should be uncommented

**Comment by dboltovskyi** [comment]
@ShobhitAd Thank you for the review.
Regarding force push we totally agree with you and understand your concerns. We've started using back merge strategy since release 6.1.

**Comment by dboltovskyi** [comment]
@ShobhitAd Please notice:
1) Conflicts were resolved in 99e0163 
2) Scripts of this PR were updated due to changes introduced by HMI PTU feature in 317ad6e 
3) Previously commented scripts (16, 17) of HMI PTU feature was included in 471917a
4) The following scripts are failed on `sdl_core` branch from [#3208](https://github.com/smartdevicelink/sdl_core/pull/3208) with `develop` back merged (without conflicts):
```
./test_scripts/Defects/6_1/3136_02_External_Proprietary.lua
./test_scripts/Policies/HMI_PTU/016_Successful_PTU_after_failed_retry_parallel_trigger_ext.lua
```
There is an odd `SDL.OnStatusUpdate(UPDATE_NEEDED)` message is sent by SDL once 2nd app is registered during retry sequence.
These 2 scripts are related to `EXTERNAL_PROPRIETARY` policy flow. Similar scripts for `PROPRIETARY` mode are passed.
Our team is going to analyze this issue and provide a fix in additional commit in [#3208](https://github.com/smartdevicelink/sdl_core/pull/3208)

**Comment by dboltovskyi** [comment]
@ShobhitAd Please notice some updates were made in scripts in last 3 commits (019287c, 9f8265f and 536a785). 
These updates are ready for review.

**Comment by dboltovskyi** [comment]
@jacobkeeler Thank for the review. 
Please notice we've just complete a few runs of full regression testing and identify some instability with 2 scrips.
It is related to timeout when applications have to be registered before PTU starts.
So small fix in 482cdb9 is applied to increase `ApplicationListUpdateTimeout` from default `2` to `4` sec.
